clang/lib/AST/ODRHash.cpp

```diff
@@ -1249,3 +1249,5 @@ void ODRHash::AddQualType(QualType T) {
 void ODRHash::AddBoolean(bool Value) {
   Bools.push_back(Value);
 }
+
+void ODRHash::AddInteger(unsigned Value) { ID.AddInteger(Value); }
```

Comment:
I remember @hahnjo and @zygoloid discussing that the odr-hasher is probably not the best way to has template arguments because the hasher would not take into account semantic aspects of template arguments. For example, a fully qualified template argument would not compare the same to a non-qualified one. We might need to implement our own folding set logic.

@hahnjo, could you help me out dig that discussion.

Reply:
Interesting. I didn't recognize this. If this is true, we need to decide if we can leave a FIXME here or we must fix it to proceed.

Reply:
The review related to `ODRHash` is this one: https://reviews.llvm.org/D153003

In short, my understanding is that `ODRHash` gives the following guarantee: If the hashes are different, there is guaranteed to be a ODR violation. In the other direction, if two hashes are the same, the declarations have to be compared in more detail, ie there may or may not be an ODR violation.

For the specializations, we need the opposite: If two template arguments are semantically the same (*), they *must* hash to the same value or otherwise we will not find the correct bucket. On the other hand, two different specialization arguments may have the same hash, that's fine for the map data structure.

Now the additional caveat (*) is that "semantically the same" is not the same congruence as "no ODR violation". In https://reviews.llvm.org/D153003 we discuss `using` declarations, but IIRC it's also possible to construct problematic cases with (nested) namespaces, top-level `::` prefixes, and template template parameters. Taken together, my conclusion from the discussion above is that `ODRHash` is simply not the right method to find template specialization parameters in a map.

Reply:
Great analysis. Fair enough, let's find a method to proceed.

Reply:
I tried to add a test case to show the problem in https://github.com/llvm/llvm-project/commit/9b808a4beb8e6c8255b412fdd6f5a3e20cbcf270. But the current patch works well for that. While I agree the ODRHash may be too aggressive for the problem we're solving, I don't want to write things that can't be well tested. I am wondering if we can proceed by leaving a FIXME here if we can't find good test in time? Or maybe we can add an option `-fload-specialization-lazily`, then we can regress smoothly if there are any problems.

@hahnjo @vgvassilev

Reply:
The secret why ODRHash can handle this may live in https://github.com/llvm/llvm-project/blob/fe1364f1e7ac0c4d0f9a4b15189485782241190d/clang/lib/AST/ODRHash.cpp#L868-L910. We've already done the work to strip the types.

Reply:
That test does not exercise an alias argument to a template template argument. IIRC the code you code is only active for typenames. Also see the test in https://reviews.llvm.org/D153003 that exercises different spellings of the semantically equivalent type `NS::A` inside a namespace.

Reply:
I had this https://github.com/llvm/llvm-project/commit/9b808a4beb8e6c8255b412fdd6f5a3e20cbcf270#diff-e88c8434346144fdf335c14c17a112759af7a4e81957efcaca635a08b9f13a29R110-R116.

Do you mean I need to put that into a namespace?

I tried to create alias arguments for template template arguments. But the tests can't pass before the current patch.

Reply:
It looks like the qualified related problems in ODRHash (at least some of them) are fixed in https://reviews.llvm.org/D156210

Reply:
I guess the comment we are discussing is here: https://reviews.llvm.org/D154324#4524368 by @zygoloid:

"
...

For [D41416](https://reviews.llvm.org/D41416), ODR hashing may not be the best mechanism to hash the template arguments, unfortunately. ODR hashing is (or perhaps, should be) about determining whether two things are spelled the same way and have the same meaning (as required by the C++ ODR), whereas I think what you're looking for is whether they have the same meaning regardless of spelling. Maybe we can get away with reusing ODR hashing anyway, on the basis that any canonical, non-dependent template argument should have the same (invented) spelling in every translation unit, but I'm not certain that's true in all cases. There may still be cases where the canonical type includes some aspect of "whatever we saw first", in which case the ODR hash can differ across translation units for non-dependent, canonical template arguments that are spelled differently but have the same meaning, though I can't think of one off-hand.
```

Reply:
Yeah, I just saw it. My concern for reinventing a new hash mechanism is how can we make sure it is correct. It may be not hard to invent a new hasher. But I am just worrying it may not be well tested. I prefer to make it step by step.

Reply:
If the example of @hahnjo works, perhaps a FIXME referring to this discussion should be sufficient and we can revisit the issue once we have an example that breaks.

Reply:
Done